File: java/jda/src/main/java/honeypot/Honeypot.java
package honeypot;

import net.dv8tion.jda.api.JDA;
import net.dv8tion.jda.api.JDABuilder;
import net.dv8tion.jda.api.OnlineStatus;
import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.entities.Activity;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.api.entities.channel.ChannelType;
import net.dv8tion.jda.api.entities.channel.middleman.GuildChannel;
import net.dv8tion.jda.api.entities.channel.middleman.MessageChannel;
import net.dv8tion.jda.api.events.channel.ChannelDeleteEvent;
import net.dv8tion.jda.api.events.guild.GuildLeaveEvent;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import net.dv8tion.jda.api.events.message.MessageReceivedEvent;
import net.dv8tion.jda.api.events.session.ReadyEvent;
import net.dv8tion.jda.api.hooks.ListenerAdapter;
import net.dv8tion.jda.api.interactions.IntegrationType;
import net.dv8tion.jda.api.interactions.InteractionContextType;
import net.dv8tion.jda.api.interactions.commands.DefaultMemberPermissions;
import net.dv8tion.jda.api.interactions.commands.OptionMapping;
import net.dv8tion.jda.api.interactions.commands.OptionType;
import net.dv8tion.jda.api.interactions.commands.build.Commands;
import net.dv8tion.jda.api.interactions.commands.build.OptionData;
import net.dv8tion.jda.api.requests.GatewayIntent;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.EnumSet;
import java.util.concurrent.TimeUnit;

public class Honeypot extends ListenerAdapter {
    private static final String TOKEN = System.getenv("DISCORD_TOKEN");
    private static final Connection DB;

    static {
        if (TOKEN == null || TOKEN.isEmpty()) {
            throw new Error("DISCORD_TOKEN is required");
        }

        // Guild honeypot config is stored in SQLite so it survives restarts.
        try {
            String dbPath = System.getenv("DB_PATH");
            if (dbPath == null) dbPath = "bot.db";
            DB = DriverManager.getConnection("jdbc:sqlite:" + dbPath);
            try (Statement stmt = DB.createStatement()) {
                stmt.execute("""
                    CREATE TABLE IF NOT EXISTS honeypots (
                      guild_id TEXT PRIMARY KEY,
                      channel_id TEXT NOT NULL,
                      action TEXT NOT NULL DEFAULT 'ban',
                      log_channel_id TEXT
                    )
                    """);
            }
        } catch (SQLException e) {
            throw new Error("Failed to open database", e);
        }
    }

    private record HoneypotConfig(String channelId, String action, String logChannelId) {}

    public static void main(String[] args) {
        Thread.setDefaultUncaughtExceptionHandler((thread, error) ->
                System.err.println("Unhandled Exception: " + error));

        // Guild messages is the only intent the honeypot needs (GUILDS is always enabled).
        JDABuilder.createDefault(TOKEN)
                .enableIntents(GatewayIntent.GUILD_MESSAGES)
                .addEventListeners(new Honeypot())
                .build();
    }

    @Override
    public void onReady(ReadyEvent event) {
        // Register slash commands once on startup.
        event.getJDA().getPresence().setStatus(OnlineStatus.ONLINE);
        event.getJDA().getPresence().setActivity(Activity.customStatus("Watching #honeypot for bots"));
        System.out.println("Logged in as " + event.getJDA().getSelfUser().getName()
                + "#" + event.getJDA().getSelfUser().getDiscriminator());
        registerCommand(event.getJDA());
    }

    @Override
    public void onSlashCommandInteraction(SlashCommandInteractionEvent event) {
        handleCommand(event);
    }

    @Override
    public void onMessageReceived(MessageReceivedEvent event) {
        handleMessage(event);
    }

    @Override
    public void onChannelDelete(ChannelDeleteEvent event) {
        // Clean up config when a channel goes away.
        deleteByChannel(event.getChannel().getId());
    }

    @Override
    public void onGuildLeave(GuildLeaveEvent event) {
        // Clean up config when the bot leaves a guild.
        deleteByGuild(event.getGuild().getId());
    }

    private static void registerCommand(JDA jda) {
        jda.updateCommands().addCommands(
                Commands.slash("honeypot-set", "Set/update honeypot channel (note: this overrides previous config set)")
                        .setDefaultPermissions(DefaultMemberPermissions.enabledFor(Permission.BAN_MEMBERS, Permission.MANAGE_CHANNEL))
                        .setContexts(InteractionContextType.GUILD)
                        .setIntegrationTypes(IntegrationType.GUILD_INSTALL)
                        .addOptions(
                                new OptionData(OptionType.CHANNEL, "channel",
                                        "The channel to ban people that message in it", true)
                                        .setChannelTypes(ChannelType.TEXT),
                                new OptionData(OptionType.STRING, "action",
                                        "The action to take when someone messages in the honeypot channel", true)
                                        .addChoice("Ban", "ban")
                                        .addChoice("Softban", "softban")
                                        .addChoice("Disabled", "disabled"),
                                new OptionData(OptionType.CHANNEL, "log_channel",
                                        "The channel to log actions in (if ommited, then it won't log anywhere)", false)
                                        .setChannelTypes(ChannelType.TEXT,
                                                ChannelType.GUILD_PUBLIC_THREAD, ChannelType.GUILD_PRIVATE_THREAD)))
                .queue();
    }

    // Handle the /honeypot-set command.
    private static void handleCommand(SlashCommandInteractionEvent event) {
        if (event.getGuild() == null) return;

        if (!event.getName().equals("honeypot-set")) {
            event.reply("Unknown command").setEphemeral(true).queue();
            return;
        }

        if (!event.getMember().hasPermission(Permission.BAN_MEMBERS, Permission.MANAGE_CHANNEL)) {
            event.reply("You don't have permission to use this command.").setEphemeral(true).queue();
            return;
        }

        String channelId = event.getOption("channel").getAsChannel().getId();
        String action = event.getOption("action").getAsString();
        OptionMapping logChannel = event.getOption("log_channel");

        if (action.equals("disabled")) {
            // "disabled" removes the config instead of updating it.
            execute("DELETE FROM honeypots WHERE guild_id = ?", event.getGuild().getId());
            event.reply("Honeypot configuration updated: Disabled honeypot for this server.").queue();
            return;
        }

        String logChannelId = logChannel == null ? null : logChannel.getAsChannel().getId();
        execute("""
                INSERT INTO honeypots (guild_id, channel_id, action, log_channel_id)
                VALUES (?, ?, ?, ?)
                ON CONFLICT(guild_id) DO UPDATE SET
                  channel_id = excluded.channel_id,
                  action = excluded.action,
                  log_channel_id = excluded.log_channel_id
                """, event.getGuild().getId(), channelId, action, logChannelId);

        event.reply("Honeypot configuration updated: Will **" + action + "** anyone who types in <#" + channelId + "> "
                + (logChannel != null ? "and log actions to <#" + logChannelId + ">" : "and won't log actions") + ".")
                .queue();
    }

    // Ban anyone who messages in the configured honeypot channel.
    private static void handleMessage(MessageReceivedEvent event) {
        if (!event.isFromGuild() || event.getAuthor().isBot()) return;

        HoneypotConfig config = loadConfig(event.getGuild().getId());
        if (config == null || !config.channelId().equals(event.getChannel().getId())) return;

        String action = config.action();
        boolean success = true;
        try {
            if (action.equals("ban")) {
                // Purge the last hour (3600s) of the user's messages.
                event.getGuild().ban(event.getAuthor(), 1, TimeUnit.HOURS)
                        .reason("User typed in #honeypot channel -> ban")
                        .complete();
            } else if (action.equals("softban")) {
                // Purge the last hour (3600s) of the user's messages.
                event.getGuild().ban(event.getAuthor(), 1, TimeUnit.HOURS)
                        .reason("User typed in #honeypot channel -> softban (1/2)")
                        .complete();
                event.getGuild().unban(event.getAuthor())
                        .reason("User typed in #honeypot channel -> softban (2/2)")
                        .complete();
            } else {
                throw new IllegalStateException("Unknown honeypot action: " + action);
            }
        } catch (RuntimeException error) {
            success = false;
            System.err.println("Failed honeypot action (" + action + "): " + error);
        }

        // On failure, log to the honeypot channel itself so moderators still see it.
        String targetChannelId = success
                ? config.logChannelId()
                : config.logChannelId() != null ? config.logChannelId() : event.getChannel().getId();
        if (targetChannelId == null) return;

        GuildChannel targetChannel = event.getGuild().getGuildChannelById(targetChannelId);
        if (!(targetChannel instanceof MessageChannel textChannel)) return;

        try {
            textChannel.sendMessage(success
                    ? "User <@" + event.getAuthor().getId() + "> was " + action
                        + " for triggering the honeypot in <#" + event.getChannel().getId() + ">"
                    : "User <@" + event.getAuthor().getId() + "> triggered the honeypot but I **failed** to " + action
                        + " them, please check my permissions to ensure I can " + action + " them.")
                    .setAllowedMentions(EnumSet.noneOf(Message.MentionType.class))
                    .complete();
        } catch (RuntimeException error) {
            System.err.println("Failed to send honeypot log message: " + error);
        }
    }

    private static synchronized HoneypotConfig loadConfig(String guildId) {
        try (PreparedStatement stmt = DB.prepareStatement(
                "SELECT channel_id, action, log_channel_id FROM honeypots WHERE guild_id = ?")) {
            stmt.setString(1, guildId);
            try (ResultSet rs = stmt.executeQuery()) {
                if (rs.next()) {
                    return new HoneypotConfig(
                            rs.getString("channel_id"),
                            rs.getString("action"),
                            rs.getString("log_channel_id"));
                }
            }
        } catch (SQLException e) {
            System.err.println("Failed to load honeypot config: " + e);
        }
        return null;
    }

    private static synchronized void execute(String sql, String... params) {
        try (PreparedStatement stmt = DB.prepareStatement(sql)) {
            for (int i = 0; i < params.length; i++) {
                stmt.setString(i + 1, params[i]);
            }
            stmt.executeUpdate();
        } catch (SQLException e) {
            System.err.println("Failed to execute query: " + e);
        }
    }

    private static void deleteByChannel(String channelId) {
        execute("DELETE FROM honeypots WHERE channel_id = ?", channelId);
    }

    private static void deleteByGuild(String guildId) {
        execute("DELETE FROM honeypots WHERE guild_id = ?", guildId);
    }
}

File: kotlin/kord/src/main/kotlin/Honeypot.kt
import dev.kord.common.entity.ChannelType
import dev.kord.common.entity.Permission
import dev.kord.common.entity.Permissions
import dev.kord.common.entity.PresenceStatus
import dev.kord.common.entity.Snowflake
import dev.kord.core.Kord
import dev.kord.core.behavior.ban
import dev.kord.core.behavior.channel.createMessage
import dev.kord.core.behavior.interaction.respondEphemeral
import dev.kord.core.behavior.interaction.respondPublic
import dev.kord.core.entity.channel.GuildMessageChannel
import dev.kord.core.entity.interaction.GuildChatInputCommandInteraction
import dev.kord.core.event.channel.ChannelDeleteEvent
import dev.kord.core.event.gateway.ReadyEvent
import dev.kord.core.event.guild.GuildDeleteEvent
import dev.kord.core.event.interaction.GuildChatInputCommandInteractionCreateEvent
import dev.kord.core.event.message.MessageCreateEvent
import dev.kord.core.on
import dev.kord.gateway.Intent
import dev.kord.rest.builder.interaction.channel
import dev.kord.rest.builder.interaction.string
import dev.kord.rest.builder.message.allowedMentions
import java.sql.Connection
import java.sql.DriverManager
import java.sql.SQLException
import java.util.concurrent.atomic.AtomicBoolean
import kotlin.time.Duration.Companion.seconds

private val TOKEN = System.getenv("DISCORD_TOKEN")
    ?.takeIf { it.isNotEmpty() }
    ?: throw Error("DISCORD_TOKEN is required")

// Guild honeypot config is stored in SQLite so it survives restarts.
private val DB: Connection = run {
    val dbPath = System.getenv("DB_PATH") ?: "bot.db"
    try {
        val connection = DriverManager.getConnection("jdbc:sqlite:$dbPath")
        connection.createStatement().use { statement ->
            statement.execute(
                """
                CREATE TABLE IF NOT EXISTS honeypots (
                  guild_id TEXT PRIMARY KEY,
                  channel_id TEXT NOT NULL,
                  action TEXT NOT NULL DEFAULT 'ban',
                  log_channel_id TEXT
                )
                """.trimIndent(),
            )
        }
        connection
    } catch (error: SQLException) {
        throw Error("Failed to open database", error)
    }
}

private val COMMANDS_REGISTERED = AtomicBoolean(false)

private data class HoneypotConfig(
    val channelId: String,
    val action: String,
    val logChannelId: String?,
)

suspend fun main() {
    val kord = Kord(TOKEN)

    kord.on<ReadyEvent> {
        // Register slash commands once on startup.
        println("Logged in as ${self.tag}")
        if (COMMANDS_REGISTERED.compareAndSet(false, true)) {
            registerCommand(kord)
        }
    }

    kord.on<GuildChatInputCommandInteractionCreateEvent> {
        handleCommand(interaction)
    }

    kord.on<MessageCreateEvent> {
        handleMessage(this)
    }

    // Clean up config when a channel or guild goes away.
    kord.on<ChannelDeleteEvent> {
        deleteByChannel(channel.id.toString())
    }
    kord.on<GuildDeleteEvent> {
        deleteByGuild(guildId.toString())
    }

    kord.login {
        presence {
            status = PresenceStatus.Online
            state = "Watching #honeypot for bots"
        }
        // Guilds + guild messages are the only intents the honeypot needs.
        intents {
            +Intent.Guilds
            +Intent.GuildMessages
        }
    }
}

private suspend fun registerCommand(kord: Kord) {
    // kord-core 0.18.1's command builder has no contexts/integrationTypes
    // options (added in later versions), so the commands can only be
    // guild-only with default (guild-install) integration types.
    kord.createGlobalApplicationCommands {
        input(
            "honeypot-set",
            "Set/update honeypot channel (note: this overrides previous config set)",
        ) {
            defaultMemberPermissions = Permissions(Permission.BanMembers, Permission.ManageChannels)
            channel("channel", "The channel to ban people that message in it") {
                required = true
                channelTypes = listOf(ChannelType.GuildText)
            }
            string("action", "The action to take when someone messages in the honeypot channel") {
                required = true
                choice("Ban", "ban")
                choice("Softban", "softban")
                choice("Disabled", "disabled")
            }
            channel("log_channel", "The channel to log actions in (if ommited, then it won't log anywhere)") {
                required = false
                channelTypes = listOf(ChannelType.GuildText, ChannelType.PublicGuildThread, ChannelType.PrivateThread)
            }
        }
    }
}

// Handle the /honeypot-set command.
private suspend fun handleCommand(interaction: GuildChatInputCommandInteraction) {
    if (interaction.command.rootName != "honeypot-set") {
        interaction.respondEphemeral { content = "Unknown command" }
        return
    }

    val required = Permissions(Permission.BanMembers, Permission.ManageChannels)
    if (required !in interaction.permissions) {
        interaction.respondEphemeral { content = "You don't have permission to use this command." }
        return
    }

    val channelId = interaction.command.channels["channel"]?.id ?: return
    val action = interaction.command.strings["action"] ?: return
    val logChannel = interaction.command.channels["log_channel"]?.id

    if (action == "disabled") {
        // "disabled" removes the config instead of updating it.
        execute("DELETE FROM honeypots WHERE guild_id = ?", interaction.guildId.toString())
        interaction.respondPublic { content = "Honeypot configuration updated: Disabled honeypot for this server." }
        return
    }

    val logChannelId = logChannel?.toString()
    execute(
        """
        INSERT INTO honeypots (guild_id, channel_id, action, log_channel_id)
        VALUES (?, ?, ?, ?)
        ON CONFLICT(guild_id) DO UPDATE SET
          channel_id = excluded.channel_id,
          action = excluded.action,
          log_channel_id = excluded.log_channel_id
        """.trimIndent(),
        interaction.guildId.toString(),
        channelId.toString(),
        action,
        logChannelId,
    )

    interaction.respondPublic {
        content = "Honeypot configuration updated: Will **$action** anyone who types in <#$channelId> " +
            (if (logChannel != null) "and log actions to <#$logChannelId>" else "and won't log actions") + "."
    }
}

// Ban anyone who messages in the configured honeypot channel.
private suspend fun handleMessage(event: MessageCreateEvent) {
    val author = event.message.author ?: return
    val guildId = event.guildId ?: return
    if (author.isBot) return

    val config = loadConfig(guildId.toString()) ?: return
    if (config.channelId != event.message.channelId.toString()) return

    val guild = event.getGuildOrNull() ?: return

    var success = true
    try {
        if (config.action == "ban") {
            guild.ban(author.id) {
                deleteMessageDuration = 3600.seconds // 1hr
                reason = "User typed in #honeypot channel -> ban"
            }
        } else if (config.action == "softban") {
            guild.ban(author.id) {
                deleteMessageDuration = 3600.seconds // 1hr
                reason = "User typed in #honeypot channel -> softban (1/2)"
            }
            guild.unban(author.id, "User typed in #honeypot channel -> softban (2/2)")
        } else {
            throw IllegalStateException("Unknown honeypot action: ${config.action}")
        }
    } catch (error: Throwable) {
        success = false
        System.err.println("Failed honeypot action (${config.action}): $error")
    }

    // On failure, log to the honeypot channel itself so moderators still see it.
    val targetChannelId = if (success) config.logChannelId else config.logChannelId ?: event.message.channelId.toString()
    if (targetChannelId == null) return

    val targetChannel = guild.getChannelOrNull(Snowflake(targetChannelId))
    if (targetChannel !is GuildMessageChannel) return

    try {
        targetChannel.createMessage {
            content = if (success) {
                "User <@${author.id}> was ${config.action} for triggering the honeypot in <#${event.message.channelId}>"
            } else {
                "User <@${author.id}> triggered the honeypot but I **failed** to ${config.action} them, " +
                    "please check my permissions to ensure I can ${config.action} them."
            }
            allowedMentions { }
        }
    } catch (error: Throwable) {
        System.err.println("Failed to send honeypot log message: $error")
    }
}

@Synchronized
private fun loadConfig(guildId: String): HoneypotConfig? {
    try {
        DB.prepareStatement("SELECT channel_id, action, log_channel_id FROM honeypots WHERE guild_id = ?").use { statement ->
            statement.setString(1, guildId)
            statement.executeQuery().use { result ->
                if (result.next()) {
                    return HoneypotConfig(
                        channelId = result.getString("channel_id"),
                        action = result.getString("action"),
                        logChannelId = result.getString("log_channel_id"),
                    )
                }
            }
        }
    } catch (error: SQLException) {
        System.err.println("Failed to load honeypot config: $error")
    }
    return null
}

@Synchronized
private fun execute(sql: String, vararg params: String?) {
    try {
        DB.prepareStatement(sql).use { statement ->
            params.forEachIndexed { index, param -> statement.setString(index + 1, param) }
            statement.executeUpdate()
        }
    } catch (error: SQLException) {
        System.err.println("Failed to execute query: $error")
    }
}

private fun deleteByChannel(channelId: String) {
    execute("DELETE FROM honeypots WHERE channel_id = ?", channelId)
}

private fun deleteByGuild(guildId: String) {
    execute("DELETE FROM honeypots WHERE guild_id = ?", guildId)
}
